File: MicroEEVEE/FollowTheWall.c
#include "Sensors.h"
#include "rmi-mr32.h"
#include <string.h>
#include <stdlib.h>

#include "FollowTheWall.h"
bool wallOnTheRight = false, wallOnTheLeft = false;
bool facedBeacon = false, keepValues=false;
double lastKnownBeaconAngle = 0, lastKnownBeaconCompass = 0;
double grausRodados = 0.0, prevComp = 0.0, prevSpeed=0, mainSide=0;
int sharpTurnCounter=-1;

void atBeacon(){
  printf("Stopped following wall!\n");
  wallOnTheRight = false; 
  wallOnTheLeft = false;
}

int minimum(int a, int b){
	if (a<b)
		return a;
	return b;
}

char* ftw_getName() {
    return "FollowTheWall";
}

void ftw_execute() {
    mainSide = obstValRight;
    double weakSide = obstValLeft, midSide = obstValFront;
    
    if (!wallOnTheRight) {
        mainSide = obstValLeft;
        weakSide = obstValRight;
    }

    led(0,0); led(1,0); led(2,0); led(3,0);
    int wallSpeed=0, opposSpeed=0;

    if ((mainSide < 10 || midSide < 5)) { 								// REALLY CLOSE WALL ON MAINSIDE OR FRONT (less than 10)
        opposSpeed = -30;        
        wallSpeed = 30;
    } else if (weakSide < 10 && mainSide > 20) {						// REALLY CLOSE WALL ON WEAKSIDE; MEDIUM WALL ON MAINSIDE (less than 10)
        opposSpeed = 30;        
        wallSpeed = -30;
    } else if ((mainSide < 15 && weakSide > 20) || midSide < 10) {  	// CLOSE WALL ON MAINSIDE OR FRONT (less than 15)
    	int dist = (minimum(mainSide, midSide)-10)*2;			// Dist between 0 and 20
    	opposSpeed = dist;
        wallSpeed = 60 + dist/2;                        
        sharpTurnCounter=-1;
    } else if (mainSide < 35 && weakSide > 25) {    					// MEDIUM WALL ON MAINSIDE (between 15 and 35) and far wall on weak side
		int dist = (mainSide-25);								// Dist between -10 and 10      
		opposSpeed = 60+dist;									// 70,50						
        wallSpeed = 60-dist;   
        prevSpeed=(mainSide-25)*2;													// prevSpeed between -20 and 20; positive means its away from wall
        sharpTurnCounter=-1;
    } else if (mainSide < 35 && weakSide <=25) {    					// MEDIUM WALL ON MAINSIDE (between 15 and 35) but medium wall on weak side as well
	   if(mainSide<weakSide-7.5){
            int dist = (mainSide-25);                           // Dist between -10 and 10
            opposSpeed = 45+dist;                               // 55,35
            wallSpeed = 45-dist;   
        } else if(weakSide<mainSide-7.5){                                 // off wall is closer!
            int dist = (weakSide-17.5)*4/3;                     // Dist between -10 and 10
            opposSpeed = 45-dist;                               // 55,35
            wallSpeed = 45+dist;                       
        } else {                                                        // walls are relatively at the same distance
            opposSpeed = 45;                               
            wallSpeed = 45;
        }
    	prevSpeed=(mainSide-25)*2;                                                  // prevSpeed between -20 and 20; positive means its away from wall	
        sharpTurnCounter=-1;											
    }
    else if (mainSide >= 35) {
      // FAR WALL ON mainSide (over 35)
        if(sharpTurnCounter<0)
            sharpTurnCounter = abs(prevSpeed)/4;                                    // turn counter between 0 and 10

    	if(prevSpeed>0){												
    	 	led(0,1); led(1,1); led(2,1); led(3,1); 
    		opposSpeed = 60;									// 60, 30-10	 curva apertada
        	wallSpeed = 35 - prevSpeed;            //35 15

        	prevSpeed=prevSpeed-3;
    	}
    	else if(prevSpeed<=0){
            led(0,0); led(1,1); led(2,1); led(3,0);
		    opposSpeed = 65;									// 65, 35-55	 Curva larga
		    wallSpeed = 35 + (-prevSpeed); 
		    prevSpeed+=3;
        }

    } else {
    	opposSpeed = 20;
        wallSpeed = 20;
    }

    if( mainSide == obstValRight){
    	setVel2(opposSpeed, wallSpeed);
    } else {
    	setVel2(wallSpeed, opposSpeed);
    }
}

void prep_ftw(){
    if (visible) {
        lastKnownBeaconAngle = beaconDir;
        lastKnownBeaconCompass = compass - lastKnownBeaconAngle;
    } else {
        lastKnownBeaconAngle = compass - lastKnownBeaconCompass;
    }

    prevComp = compass;
}

bool ftw_isPossible() {
    if(sharpTurnCounter>=0)
        sharpTurnCounter--;

    // Beacon visible, on the right side, with a wall on the right
    bool rightWallOnBeaconSide = visible && obstValRight < 20 && lastKnownBeaconAngle >= 0 && lastKnownBeaconAngle <= 60;
    bool leftWallOnBeaconSide = visible && obstValLeft < 20 && lastKnownBeaconAngle <= 0 && lastKnownBeaconAngle >= -60;
    bool beaconNotVisWithWall = !visible && (obstValLeft < 20 || obstValRight < 20);

    //if(!wallOnTheRight && !wallOnTheLeft && (rightWallOnBeaconSide || leftWallOnBeaconSide || beaconNotVisWithWall))
		//printf("Gonna start following wall: rightW:%d leftW:%d notVisWall:%d\n",rightWallOnBeaconSide, leftWallOnBeaconSide,beaconNotVisWithWall);
    
    double distanceNeededToAbandonWall=0;
    if(abs(lastKnownBeaconAngle)<10)
    	distanceNeededToAbandonWall=15;
    else
    	distanceNeededToAbandonWall=(abs(lastKnownBeaconAngle)-10)*0.4+15;

    // IF FOLLOWING WALL , beacon wasn't visible but now is
    // podemos mudar isto e po-lo a largar a parede caso deixe d sentir uma parede e nao souber nunca do beacon
    if (!facedBeacon && visible && (wallOnTheRight || wallOnTheLeft) && sharpTurnCounter<=0 ) {
        wallOnTheRight = false;
        wallOnTheLeft = false;
    } 
    // if following wall, beacon has been visible all along, beacon is in front, no obstacles nearby and not in any loops
    else if (facedBeacon && visible && (wallOnTheRight || wallOnTheLeft) && (lastKnownBeaconAngle < 60 && lastKnownBeaconAngle > -60) 
    	&& obstValFront > 15 && mainSide > distanceNeededToAbandonWall && sharpTurnCounter<=0 ){ //&& grausRodados < 180 && grausRodados > -180) {
        wallOnTheRight = false;
        wallOnTheLeft = false;
    } 
    // IF NOT FOLLOWING WALL and near a wall
    else if ((!wallOnTheRight && !wallOnTheLeft) && (rightWallOnBeaconSide || leftWallOnBeaconSide || beaconNotVisWithWall)) {
        wallOnTheLeft = obstValLeft < obstValRight;
        wallOnTheRight = !wallOnTheLeft;

        if(!keepValues){
        	prevSpeed = 0;
        } else {
        	keepValues = 0;
        }
        facedBeacon = visible;
        grausRodados = 0;

    }

    return wallOnTheRight || wallOnTheLeft;
}

File: Legacy/PathFinderD/Robotics1/src/Engine/FollowTheWall.java
package Engine;

import static Engine.Repetitions.*;
import static Engine.Sensors.*;
import ciberIF.ciberIF;

public class FollowTheWall implements Behaviour {

    boolean wallOnTheRight = false, wallOnTheLeft = false, facedBeacon = false, startFollowingWithoutBeacon = true;
    double lastKnownBeaconAngle = 0, lastKnownBeaconCompass = 0;
    double grausRodados = 0.0, prevComp = 0.0;

    public FollowTheWall(boolean fwa) {
        startFollowingWithoutBeacon = fwa;
    }

    @Override
    public void execute(ciberIF cif) {
        //System.out.printf("\tFeeling middle=%2.1f left=%2.1f right=%2.1f\n", irSensor0, irSensor1, irSensor2);

        double modder = 1.0;
        double mainSide = irSensor2, weakSide = irSensor1, midSide = irSensor0;
        if (!wallOnTheRight) {
            modder = -1.0;
            mainSide = irSensor1;
            weakSide = irSensor2;
        }

        /*boolean avoider = avoidRepetitions();
        if (avoider) {
            System.out.println("AVOID!");
        }*/

        //need to change 1 and 0 to modder influenced
        if ( (mainSide > 4.0 || midSide > 4.0)) { //avoider ||
            cif.DriveMotors(modder * -0.1, modder * 0.1);
            history[counter] = getMainSideTurning(modder);
        } else if (weakSide > 4.0 && mainSide < 2.0) {
            cif.DriveMotors(modder * 0.1, modder * -0.1);
            history[counter] = getWeakSideTurning(modder);
        } else if ((mainSide >= 3.0 && weakSide < 2.0) || midSide > 2.0) {  // if close wall on the right OR close wall in front
            cif.DriveMotors(modder * -0.1, modder * 0.1);   // sharp turn left
            history[counter] = getMainSideTurning(modder);
        } else if (mainSide > 2.5) {                         // if far wall on the right and nothing in front
            cif.DriveMotors(0.09 - modder * 0.01, 0.9 + modder * 0.01);    //  straight forward
            history[counter] = -1;
        } else if (mainSide > 2.0) {                         // if far wall on the right and nothing in front
            cif.DriveMotors(0.1, 0.1);    //  straight forward
            history[counter] = -1;
        } else if (mainSide > 1.0) {                         // turn slightly right
            cif.DriveMotors(0.06 + (modder * 0.04), 0.06 - (modder * 0.04));
            history[counter] = -1;
        } else if (mainSide <= 1.0) {                        // if wall has vanished
            cif.DriveMotors(0.05 + (modder * 0.05), 0.05 - (modder * 0.05));    //  turn right
            history[counter] = -1;
        }

        counter = (counter + 1) % history.length;
    }

    @Override
    public boolean isPossible() {
        double modder = 0;
        if (compass < -150 && prevComp > 150) {
            modder = 360 - Math.abs(compass) - Math.abs(prevComp);
        } else if (compass > 150 && prevComp < -150) {
            modder = -(360 - Math.abs(compass) - Math.abs(prevComp));
        } else {
            modder = compass - prevComp;
        }
        grausRodados += modder;

        if (beacon.beaconVisible) {
            lastKnownBeaconAngle = beacon.beaconDir;
            lastKnownBeaconCompass = compass - lastKnownBeaconAngle;
        } else {
            lastKnownBeaconAngle = compass - lastKnownBeaconCompass;
        }

        //System.out.printf("\t ------> Compass: %2.1f    Beacon: %2.1f     Graus: %2.1f\n", compass, lastKnownBeaconAngle, grausRodados);
        
        // IF FOLLOWING WALL 
        if (!facedBeacon && beacon.beaconVisible && (wallOnTheRight || wallOnTheLeft)) {
            wallOnTheRight = false;
            wallOnTheLeft = false;
        } else if (facedBeacon && (wallOnTheRight || wallOnTheLeft)) {
            // if beacon in front and no obstacles
            if ((lastKnownBeaconAngle < 60 && lastKnownBeaconAngle > -60) && irSensor0 < 1.0 && grausRodados < 180 && grausRodados > -180) {//distance < 100) {
                wallOnTheRight = false;
                wallOnTheLeft = false;
            }
            // IF NOT FOLLOWING WALL and facing beacon
        } else if (!wallOnTheRight && !wallOnTheLeft && (startFollowingWithoutBeacon || (beacon.beaconVisible && lastKnownBeaconAngle < 60 && lastKnownBeaconAngle > -60))) {

            if (irSensor2 > 2.0 || irSensor1 > 2.0) {     //|| irSensor0 > 1.0    // if beacon in front and obstacles as well
                //System.out.println("LeftSensor= "+irSensor1+ "    RightSensor="+irSensor2);
                wallOnTheLeft = irSensor1 > irSensor2;
                wallOnTheRight = !wallOnTheLeft;

                facedBeacon = beacon.beaconVisible;

                //distance = 100;
                //inertia = 0;
                //anguloInicial = compass;
                grausRodados = 0;
            }
        }

        prevComp = compass;

        return wallOnTheRight || wallOnTheLeft;
    }

    @Override
    public String getName() {
        return "FollowTheWall";
    }

}
